core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -303,17 +306,19 @@ public void run() {
       return jumpListener;
     }
 
-    private Future<?> scheduleTimeout(final ServerStream stream, Metadata headers) {
+    private Future<?> scheduleTimeout(final ServerStream stream, Metadata headers,
+                                      final Context.CancellableContext context) {
       Long timeoutMicros = headers.get(TIMEOUT_KEY);
       if (timeoutMicros == null) {
         return DEFAULT_TIMEOUT_FUTURE;
       }
-      return timeoutService.schedule(new Runnable() {
+      return timeoutService.schedule(new ContextRunnable(context) {
           @Override
-          public void run() {
+          public void runInContext() {
             // This should rarely get run, since the client will likely cancel the stream before
             // the timeout is reached.
             stream.cancel(Status.DEADLINE_EXCEEDED);
+            context.cancel(null);
```

Comment:
java.util.concurrent.TimeoutException since that was what was done in Context? That seems like an exception we would want to suppress filling in the stack trace.